R/data.table.R

```diff
@@ -1739,41 +1739,79 @@ replace_dot_alias = function(e) {
         GForce = FALSE
       } else {
         # Apply GForce
+        # GForce needs to evaluate all arguments not present in the data.table before calling C part #5547
+        # Safe cases: variables [i], calls without variables [c(0,1), list(1)] # TODO extend this list
+        # Unsafe cases: functions containing variables [c(i), abs(i)], .N
+        is_constantish = function(expr, check) {
+          (!is.call(expr) || length(all.vars(expr, max.names=1L, unique=FALSE))==0L) && 
+            !dotN(expr) &&
+            (missing(check) || check(expr))
```

Comment:
```suggestion
        is_constantish = function(expr, check_singleton) {
          if (!is.call(expr)) {
            return(!dotN(expr))
          }
          if (check_singleton) {
            return(FALSE)
          }
          # calls are allowed <=> there's no SYMBOLs in the sub-AST
          return(length(all.vars(expr, max.names=1L, unique=FALSE))==0L)
```

I might be misunderstanding, but I think the `length=1` check is equivalent to forcing non-call, right?

So maybe this body is simpler to read

Reply:
Sounds good, change and add default value for `check_singleton=FALSE`